### PR TITLE
[ECR-4133] Add service resume

### PR DESCRIPTION
**Pull request by bullet-tooth**
## Overview

---
See: https://jira.bf.local/browse/ECR-4133

### Definition of Done

- [ ] There are no TODOs left in the code
- [ ] Change is covered by automated [tests](https://github.com/exonum/exonum-java-binding/blob/master/CONTRIBUTING.md#tests)
- [ ] The [coding guidelines](https://github.com/exonum/exonum-java-binding/blob/master/CONTRIBUTING.md#the-code-style) are followed
- [ ] Public API has Javadoc
- [ ] Method preconditions are checked and documented in the Javadoc of the method
- [ ] Changelog is updated if needed (in case of notable or breaking changes)
- [ ] The continuous integration build passes

### CODE REVIEW

File: exonum-java-binding/core/src/main/java/com/exonum/binding/core/runtime/ServiceRuntimeAdapter.java [hunk 1]
```diff
@@ -151,6 +151,20 @@ void initiateAddingService(long forkHandle, byte[] instanceSpec, byte[] configur
...
+  /**
+   * Starts resuming a service with the given specification.
+   *
+   * @param instanceSpec the service instance specification as a serialized {@link InstanceSpec}
+   *     protobuf message
+   * @param configuration the service initial configuration parameters as a serialized protobuf
+   *     message
+   * @see ServiceRuntime#initializeResumingService(ServiceInstanceSpec, byte[])
+   */
+  void initializeResumingService(byte[] instanceSpec, byte[] configuration) {
```
Comment (bullet-tooth) [inline exonum-java-binding/core/src/main/java/com/exonum/binding/core/runtime/ServiceRuntimeAdapter.java]: @vitvakatu method signature for ECR-4134

File: exonum-java-binding/CHANGELOG.md [hunk 2]
```diff
@@ -36,7 +36,7 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
 - `supervisor-mode` CLI parameter added for `generate-template` command. It
   allows to configure the mode of the Supervisor service. Possible values are
   "simple" and "decentralized". (#1361)
-- Service instances can be stopped now. (#1358)
+- Service instances can be stopped and resumed now. (#1358, #1372)
```
Comment (dmitry-timofeev) [inline exonum-java-binding/CHANGELOG.md]: An instance artifact may be upgraded before resuming, but with no data changes. 

(may also insert the uses cases from the epic/ECR-3773).

File: exonum-java-binding/core/src/main/java/com/exonum/binding/core/service/Service.java [hunk 3]
```diff
@@ -53,6 +53,24 @@ default void initialize(Fork fork, Configuration configuration) {
     // No configuration
   }
 
+  /**
+   * Performs resuming of previously stopped service instance. This method is called <em>once</em>
```
Comment (dmitry-timofeev) [inline exonum-java-binding/core/src/main/java/com/exonum/binding/core/service/Service.java]:
```
   * Resumes the previously stopped service instance. This method is called when
   * a stopped service instance is restarted.
```

File: exonum-java-binding/core/src/main/java/com/exonum/binding/core/service/Service.java [hunk 4]
```diff
@@ -53,6 +53,24 @@ default void initialize(Fork fork, Configuration configuration) {
...
+   * due to violation migration procedure.</em>
+   *
+   * @param configuration the service configuration parameters
+   * @throws ExecutionException if the configuration parameters are not valid (e.g.,
+   *     malformed, or do not meet the preconditions). Exonum will save the error into
+   *     {@linkplain com.exonum.binding.core.blockchain.Blockchain#getCallErrors(long)
+   *     the registry of call errors} if the resuming fails
+   * @see Configurable
+   */
+  default void resume(Configuration configuration) {
```
Comment (dmitry-timofeev) [inline exonum-java-binding/core/src/main/java/com/exonum/binding/core/service/Service.java]: @slowli, @bullet-tooth  In which cases the service needs to be notified of that, if it can't do anything with this configuration, just look at them? I remember that the service instance _objects_ are not supposed to have much state, therefore,

> "As an example, arguments can be used to update the service configuration."

does not seem to make much sense — that configuration must be normally kept in the persistent storage. On top of that, we do *not* provide a similar operation for _node restarts_, when a fresh service instance object is also created.

File: exonum-java-binding/core/src/main/java/com/exonum/binding/core/service/Service.java [hunk 5]
```diff
@@ -53,6 +53,24 @@ default void initialize(Fork fork, Configuration configuration) {
...
+  /**
+   * Performs resuming of previously stopped service instance. This method is called <em>once</em>
+   * after restarting a service instance.
+   *
+   * <p><em>Note that changing the data during resume operations is not allowed
+   * due to violation migration procedure.</em>
+   *
+   * @param configuration the service configuration parameters
+   * @throws ExecutionException if the configuration parameters are not valid (e.g.,
+   *     malformed, or do not meet the preconditions). Exonum will save the error into
```
Comment (dmitry-timofeev) [inline exonum-java-binding/core/src/main/java/com/exonum/binding/core/service/Service.java]: It will not save them.

File: exonum-java-binding/core/src/main/java/com/exonum/binding/core/service/Service.java [hunk 6]
```diff
@@ -53,6 +53,24 @@ default void initialize(Fork fork, Configuration configuration) {
...
+   *     {@linkplain com.exonum.binding.core.blockchain.Blockchain#getCallErrors(long)
+   *     the registry of call errors} if the resuming fails
+   * @see Configurable
```
Comment (dmitry-timofeev) [inline exonum-java-binding/core/src/main/java/com/exonum/binding/core/service/Service.java]: Configurable is not related to this, but initialize (if we decide to keep the method at all)

File: exonum-java-binding/core/src/main/java/com/exonum/binding/core/runtime/ServiceRuntime.java [hunk 7]
```diff
@@ -206,6 +206,23 @@ public void initiateAddingService(Fork fork, ServiceInstanceSpec instanceSpec,
...
+   *     message
+   * @throws IllegalArgumentException if the given service instance is active
+   */
+  public void initializeResumingService(ServiceInstanceSpec instanceSpec,
+      byte[] configuration) {
+    synchronized (lock) {
+      checkStoppedService(instanceSpec.getId());
+      ServiceWrapper service = createServiceInstance(instanceSpec);
+      service.resume(new ServiceConfiguration(configuration));
+    }
```
Comment (dmitry-timofeev) [inline exonum-java-binding/core/src/main/java/com/exonum/binding/core/runtime/ServiceRuntime.java]: Currently this class is responsible to log errors, though there is ECR-3813 to shift that.

File: exonum-java-binding/core/src/main/java/com/exonum/binding/core/runtime/ServiceRuntime.java [hunk 8]
```diff
@@ -486,16 +503,22 @@ public void close() throws InterruptedException {
...
-  private void checkService(Integer serviceId) {
+  private void checkActiveService(Integer serviceId) {
     checkArgument(servicesById.containsKey(serviceId),
         "No service with id=%s in the Java runtime", serviceId);
   }
 
+  /** Checks that the service with the given id is not active in this runtime. */
+  private void checkStoppedService(Integer serviceId) {
+    checkArgument(!servicesById.containsKey(serviceId),
+        "Service with id=%s should be stopped, but actually active", serviceId);
```
Comment (dmitry-timofeev) [inline exonum-java-binding/core/src/main/java/com/exonum/binding/core/runtime/ServiceRuntime.java]: Add the offending instance?

File: exonum-java-binding/core/src/test/java/com/exonum/binding/core/runtime/ServiceRuntimeIntegrationTest.java [hunk 9]
```diff
@@ -96,8 +96,6 @@ void startsServerOnInitialization() {
 
   @Test
   void deployCorrectArtifact() throws Exception {
-    serviceRuntime.initialize(mock(Node.class));
```
Comment (dmitry-timofeev) [inline exonum-java-binding/core/src/test/java/com/exonum/binding/core/runtime/ServiceRuntimeIntegrationTest.java]: (Here and elsewhere) Why is this removed? No operation shall occur before SR is initialized.

File: exonum-java-binding/core/src/test/java/com/exonum/binding/core/runtime/ServiceRuntimeIntegrationTest.java [hunk 10]
```diff
@@ -191,9 +183,61 @@ void startAddingService() {
...
+    serviceRuntime.initializeResumingService(instanceSpec, configuration);
+
+    // Check it was instantiated as expected
+    verify(servicesFactory).createService(eq(serviceDefinition), eq(instanceSpec),
+        any(MultiplexingNodeDecorator.class));
+
+    // and the service was configured
+    Configuration expectedConfig = new ServiceConfiguration(configuration);
+    verify(serviceWrapper).resume(expectedConfig);
+  }
```
Comment (dmitry-timofeev) [inline exonum-java-binding/core/src/test/java/com/exonum/binding/core/runtime/ServiceRuntimeIntegrationTest.java]: // but not registered in the runtime yet:
assertThat(serviceRuntime.findService(TEST_NAME)).isEmpty();

File: exonum-java-binding/core/src/test/java/com/exonum/binding/core/runtime/ServiceRuntimeIntegrationTest.java (hunk 9 above)
Comment (bullet-tooth) [reply exonum-java-binding/core/src/test/java/com/exonum/binding/core/runtime/ServiceRuntimeIntegrationTest.java]: ok, I will add a null check to the decorator then

Comment (dmitry-timofeev) [reply exonum-java-binding/core/src/test/java/com/exonum/binding/core/runtime/ServiceRuntimeIntegrationTest.java]: Not sure I understand — the SR expects that it is first initialized with a non-null Node before any other operations are performed.

File: exonum-java-binding/CHANGELOG.md [hunk 11]
```diff
@@ -36,7 +36,9 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
 - `supervisor-mode` CLI parameter added for `generate-template` command. It
   allows to configure the mode of the Supervisor service. Possible values are
   "simple" and "decentralized". (#1361)
-- Service instances can be stopped now. (#1358)
+- Support of service instances lifecycle: they can be activated, stopped and resumed now.
+  Also, service instance artifacts can be upgraded before resuming (but with no data changes)
```
Comment (dmitry-timofeev) [inline exonum-java-binding/CHANGELOG.md]: No data changes, according to the recent discussions, is no longer correct.

I'd add to the list below 'synchronous data migration'.

File: exonum-java-binding/core/src/main/java/com/exonum/binding/core/service/Service.java [hunk 12]
```diff
@@ -53,6 +53,26 @@ default void initialize(Fork fork, Configuration configuration) {
...
 
+  /**
+   * Resumes the previously stopped service instance. This method is called when
+   * a stopped service instance is restarted.
+   *
+   * <p>This method may perform any changes to the database. For example, update some service
+   * parameters, deprecate old entries etc.
+   *
+   * <p>Also, note that performing any bulk operations or data migration
+   * <em>is not recommended</em> here.
```
Comment (dmitry-timofeev) [inline exonum-java-binding/core/src/main/java/com/exonum/binding/core/service/Service.java]: I'd add short motivation, as Ostrovski clarified in the discussion.

File: exonum-java-binding/core/src/main/java/com/exonum/binding/core/service/Service.java [hunk 13]
```diff
@@ -53,6 +53,26 @@ default void initialize(Fork fork, Configuration configuration) {
...
+   * <!--TODO: Add a link to the migration procedure -->
+   *
+   * @param fork a database fork to apply changes to. Not valid after this method returns
+   * @param configuration the service configuration parameters
```
Comment (dmitry-timofeev) [inline exonum-java-binding/core/src/main/java/com/exonum/binding/core/service/Service.java]: As we've discussed, those are hardly _configuration parameters_, just generic 'arguments'.

File: exonum-java-binding/core/src/test/java/com/exonum/binding/core/runtime/ServiceRuntimeIntegrationTest.java [hunk 14]
```diff
@@ -369,6 +367,68 @@ void updateServiceStatusBadStatus(Status badStatus) {
...
+    // Create the service from the artifact
+    Fork fork = mock(Fork.class);
+    byte[] configuration = anyConfiguration();
+    serviceRuntime.initializeResumingService(fork, instanceSpec, configuration);
+
+    // Check it was instantiated as expected
+    verify(servicesFactory).createService(eq(serviceDefinition), eq(instanceSpec),
+        any(MultiplexingNodeDecorator.class));
+
+    // and the service was configured
```
Comment (dmitry-timofeev) [inline exonum-java-binding/core/src/test/java/com/exonum/binding/core/runtime/ServiceRuntimeIntegrationTest.java]: resumed

File: exonum-java-binding/core/src/main/java/com/exonum/binding/core/service/Service.java (hunk 13 above)
Comment (dmitry-timofeev) [reply exonum-java-binding/core/src/main/java/com/exonum/binding/core/service/Service.java]: `byte[]` since is it is not configuration. We shall not use Configuration for this operation — it is for configuration parameters.

It could be made one, it is made in this way probably to give more flexibility in terms of implementation (e.g., if we went with also properties).

File: exonum-java-binding/core/src/main/java/com/exonum/binding/core/runtime/ServiceRuntime.java [hunk 15]
```diff
@@ -206,6 +206,38 @@ public void initiateAddingService(Fork fork, ServiceInstanceSpec instanceSpec,
     }
   }
 
+  /**
+   * Initiates resuming of previously stopped service instance. Service instance artifact could
+   * be upgraded in advance to bring some new functionality.
+   *
+   * @param fork a database view to apply arguments
```
Comment (dmitry-timofeev) [inline exonum-java-binding/core/src/main/java/com/exonum/binding/core/runtime/ServiceRuntime.java]: _Arguments_ are not applied to a database.
